AgileFreaks OOP Test for Madalin Balint

MadalinBalint opened the pull request.

MadalinBalint committed:
Resolved problem for Coffee Shops

MadalinBalint committed:
Added JUnit tests

pirvudoru commented on src/com/madalinb/coffeeshop/csv/CSVParser.kt:
```diff
@@ -0,0 +1,31 @@
+            } else
+                println("Error: not a valid name in line '$line'")
+        } else
+            println("Error: wrong number of parameters in line '$line'")
```
What would it take to use this class in a web-based application instead of a console one?

MadalinBalint replied:
Use a different method to show the error messages (Console.log) or better throw them as an Exception.

pirvudoru commented on src/com/madalinb/coffeeshop/abstractions/validations/ValidString.kt:
```diff
@@ -0,0 +1,5 @@
+package com.madalinb.coffeeshop.abstractions.validations
+
+interface ValidString {
```
By reading the name, I would think there will be a class that inherits from `String` and implements `ValidString`.

MadalinBalint replied:
I've changed the name to a more appropriate one.

pirvudoru commented on src/com/madalinb/coffeeshop/data/CSVFileOps.kt:
```diff
@@ -0,0 +1,35 @@
+import java.io.FileNotFoundException
+import java.io.IOException
+
+class CSVFileOps(var parser: Parser<CoffeeShop>) : FileOperations<CoffeeShop>() {
```
What does "CSVFileOps" stand for and the purpose of it?

MadalinBalint replied:
File operations for the CSV file format. I've changed the name to a more appropriate one.

pirvudoru commented on .idea/libraries/junit_4_12.xml:
```diff
@@ -0,0 +1,9 @@
+<component name="libraryTable">
```
Why do we need `.idea` files added to VCS?

MadalinBalint replied:
They were added by mistake. Any IDE related configuration files shouldn't be commited to the repository.

pirvudoru commented on src/com/madalinb/coffeeshop/Main.kt:
```diff
@@ -0,0 +1,69 @@
+    coffeeShops.forEach {
+        val distance = it.distanceTo(x, y)
+        map[distance] = it.name
+    }
```
What can we do something about this logic if we want to re-use it in another context (web) ?

MadalinBalint replied:
Move the variable declaration outside of the loop.

pirvudoru commented on src/com/madalinb/coffeeshop/data/CoffeeShop.kt:
```diff
@@ -0,0 +1,7 @@
+package com.madalinb.coffeeshop.data
+
+data class CoffeeShop(var name: String, var x: Double, var y: Double) {
+    fun distanceTo(x2: Double, y2: Double): Double {
```
Why does the method accept x and y? What alternatives do we have and how does it compare to this one?

MadalinBalint replied:
In this case, because the documentation says that we have to assume that all coordinates lie on a plane, we are talking about Cartesian coordinates. In a real world where the Earth is not flat (as some people really belive), but round, the coordinates should be latitude and longitude, and the formula for calculating the distance would be different.

alexandru-calinoiu replied:
I think @pirvudoru is refering here to a code smell called [primitive obsesion](https://refactoring.guru/smells/primitive-obsession) you are inferring that 2 doubles are coordinates, but we are asking if there is a better way to do this.

MadalinBalint replied:
Moved the function into a more appropriate class.

pirvudoru commented on src/com/madalinb/coffeeshop/tests/CoffeeShopTest.kt:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Test
+    fun getName() {
```
By looking at the name of the test, i don't have much context of what it actually verifies. 

https://stackoverflow.com/questions/155436/unit-test-naming-best-practices

MadalinBalint replied:
Changed the name to a more appropriate one.

pirvudoru reviewed (commented):
Looks good! Have a few questions / suggestions that i'd like to discuss.

MadalinBalint committed:
Changed the name of some interfaces,methods,classes to a more appropriate one

MadalinBalint committed:
Small optimization

MadalinBalint committed:
Variable name correction

alexandru-calinoiu commented on src/com/madalinb/coffeeshop/Main.kt:
```diff
@@ -0,0 +1,70 @@
+    val y: Double? = args[1].toDoubleOrNull()
+    val filename = args[2]
+
+    if (args.size != 3) {
```
Isn't this check to late, as you already have assigned the params in the above lines.

MadalinBalint replied:
Moved the declarations after the check for the 3 arguments.

alexandru-calinoiu commented on src/com/madalinb/coffeeshop/csv/CSVParser.kt:
```diff
@@ -0,0 +1,31 @@
+                if (validator.isValidDouble(params[1]) &&
+                    validator.isValidDouble(params[2])
+                ) {
+                    return CoffeeShop(
```
This reminds me of https://www.audero.it/blog/wp-content/uploads/2016/08/php-street-fighter.jpg

I think this validator has a very large api surface, it should simply state if a line is valid or not.

Can you think of a better way to do this?

MadalinBalint replied:
Refactored the class, and introduced Exceptions.

alexandru-calinoiu commented on src/com/madalinb/coffeeshop/data/CoffeeShop.kt:
```diff
@@ -0,0 +1,7 @@
+package com.madalinb.coffeeshop.data
+
+data class CoffeeShop(var name: String, var x: Double, var y: Double) {
```
Usually data classes are meant to store only data and no functions, can you think of why is the case?

MadalinBalint replied:
Refactored the class.

alexandru-calinoiu commented on src/com/madalinb/coffeeshop/Main.kt:
```diff
@@ -0,0 +1,70 @@
+    var distance:Double
+    coffeeShops.forEach {
+        distance = it.distanceTo(x, y)
+        map[distance] = it.name
```
This might not be correct, if you use a map and you have 2 shops at the same distance you will only show one to the user.

MadalinBalint replied:
Changed to List.

alexandru-calinoiu commented on src/com/madalinb/coffeeshop/Main.kt:
```diff
@@ -0,0 +1,70 @@
+    val sortedMap = map.toSortedMap()
+    var counter = 0
+
+    for ((dist, name) in sortedMap) {
```
Preaty sure there is a easier way to get the first 3 elements in kotlin

MadalinBalint replied:
Refactored in class CoffeeShopFinder as function findClosest3Shops.

alexandru-calinoiu reviewed (commented):
Good job,

I like the fact that there where no dependencies on concrete classes, this makes the implementation easy to test.

Please take a few minutes to go over my comments and answer them.

MadalinBalint committed:
Added Exception for CSVParser.
Moved some operations from Main into separate functions of class CoffeeShopFinder.

alexandru-calinoiu commented:
Thank you, will close this and followup via email.
